### ISSUE TITLE
Possibility to search within all opened tabs

### ISSUE THREAD

**Issue by Keagel** [mbnuqw/sidebery#168]
For example, it could be done the same way as in "Tree Tabs" where you have a magnifying glass button which when you click lets you type into a text field, filtering in real-time.
Thank you!

**Comment by mbnuqw** [comment]
Yeah, I'll add search anyway, just a bit later (after the tags for tabs/bookmarks feature).

**Comment by ffabbroc** [comment]
I just recently switched from Power Tabs to Sideberry and do miss the Power Tabs feature of searching across all tabs.

**Comment by AneesAmjad** [comment]
> 
> 
> Yeah, I'll add search anyway, just a bit later (after the tags for tabs/bookmarks feature).

And please when you add this feature make it for keyboard ninjas. What I mean is that feature should be activated by keyboard shortcut and search it immediately without mouse usage.

**Comment by mbnuqw** [comment]
`search it immediately without mouse usage`
Oh, this is the main reason why I didn't  implemented this feature from the beginning - it can't be done using the WebExtentions API :crying_cat_face: . So the fastest way to search (for me) is using the native Firefox url bar with its [pre|post]fixes (%, *, etc).

**Comment by AneesAmjad** [comment]
> it can't be done using the WebExtentions API 😿

Well, currently I'm using this one, this can be used completely by keyboard. https://addons.mozilla.org/en-US/firefox/addon/tab_search/ 

The problem with pre|post fixes is they just show only name of the tab. Suppose if I have open multiple tabs of same websites only the website name will be visible or I have to be little more specifc by adding some url word. But with the extension you can see url which is handy. Usually dashboards like Wordpress, or FB manager have same names.

**Comment by mbnuqw** [comment]
`Well, currently I'm using this one, this can be used completely by keyboard. https://addons.mozilla.org/en-US/firefox/addon/tab_search/`
Thank you for showing me this! Looks like this is possible with popup window (it catch the focus on opening), but not in the sidebar which is usually opened all the time. I think, I can create popup with only input field and maybe some options which will communicate with the sidebar for showing results.

**Comment by AneesAmjad** [comment]
> 
> 
> `Well, currently I'm using this one, this can be used completely by keyboard. https://addons.mozilla.org/en-US/firefox/addon/tab_search/`
> Thank you for showing me this! Looks like this is possible with popup window (it catch the focus on opening), but not in the sidebar which is usually opened all the time. I think, I can create popup with only input field and maybe some options which will communicate with the sidebar for showing results.

 More power to you.  :-)

**Comment by btglr** [comment]
> I can create popup with only input field and maybe some options which will communicate with the sidebar for showing results.

I suggest you make it pop up over the center of the page (if possible), something like the ueli one? So it would look like the screenshot, and for example when you press Return it either highlights the matches or hides anything that doesn't match (only as a visual filter, maybe CSS-based?). Maybe with a button for regular expressions, that kind of thing (added it to the screenshot).

![example](https://user-images.githubusercontent.com/37577064/77926933-03f95c00-72a7-11ea-8840-8d8dc7395f6d.png)

**Comment by aletaziar** [comment]
I would suggest adding ways to filter the tabs not just by title or url, but also by status: muted, playing audio, unread, discarded, parent, etc.

**Comment by mbnuqw** [comment]
@btglr, to achieve this I need to inject a keylogger script to every active page, which is not very convenient solution. But I'll experiment with it.

@aletaziar, filtering by muted, audible, unread and discarded flags is quite simple to implement. Showing the parents chain of filtered tab possible too (to represent tree structure). But filtering by the parent tab is not clear to me. How it should look.

**Comment by aletaziar** [comment]
> @aletaziar, filtering by muted, audible, unread and discarded flags is quite simple to implement. Showing the parents chain of filtered tab possible too (to represent tree structure). But filtering by the parent tab is not clear to me. How it should look.

Be able to search only for tabs that are parent tabs is what I mean. But showing a search result's parent tab would also be an amazing feature that I didn't even think of.

> to achieve this I need to **inject a keylogger script to every active page**, which is not very convenient solution. But I'll experiment with it.

The active page injection is why I would much rather the search was executed from either the sidebar or the toolbar button. I think injection into the active page should be avoided whenever possible.

**Comment by Keagel** [comment]
> @btglr, to achieve this I need to inject a keylogger script to every active page, which is not very convenient solution. But I'll experiment with it.

I don't have much experience with browser extensions, but isn't it possible to display the search input in the middle of the screen even if it's part of the sidebar? This would seem like something that can be done with relative ease with a bit of CSS.

**Comment by mbnuqw** [comment]
@aletaziar, `injection into the active page should be avoided whenever possible`
True.

@Keagel , this is not possible b/c each part of an extension (sidebar, popups) are isolated documents.

Also there is no possibility to programmatically focus on the input element in the sidebar (e.g. using shortcut). So I think to add two options for search:
- Show search-bar in the sidebar - User will be able to click on it and start the searching process.
- Keybinding for opening popup - This will allow user, without mouse, open popup with the focused input element.

**Comment by gsbabil** [comment]
Hi Everyone, @mbnuqw, 

I tried the [Tab Search](https://addons.mozilla.org/en-US/firefox/addon/tab_search/) recommendation above. Thank you for sharing. I've also been using [Tridactyl](https://addons.mozilla.org/en-US/firefox/addon/tridactyl-vim/?src=external-github) for a while, which has the `b` shortcut for `buffer/tab` commands -- `b` shows a list of tabs and allows navigation through the `TAB` key.

One item I'd highlight here is that neither [Tab Search](https://addons.mozilla.org/en-US/firefox/addon/tab_search/) nor [Tridactyl](https://addons.mozilla.org/en-US/firefox/addon/tridactyl-vim/?src=external-github) honors the panel/group organizations in Sidebery. 

For example, if I navigated to another tab through either of the above extensions, Sidebery's sidebar panel remains unnotified and doesn't reflect any tab-navigation related changes.

Thank you.

**Comment by mbnuqw** [comment]
>@gsbabil
For example, if I navigated to another tab through either of the above extensions, Sidebery's sidebar panel remains unnotified and doesn't reflect any tab-navigation related changes.

I cannot reproduce this. Sidebery is highlighting the active tab and switch the panel if needed:
![Peek 2020-05-10 21-34](https://user-images.githubusercontent.com/6276694/81505197-12578080-9307-11ea-94fd-9fb304f46fb7.gif)

Please, open the new issue with debug info, which I could use to investigate this.

**Comment by gsbabil** [comment]
@mbnuqw 

> Please, open the new issue with debug info, which I could use to investigate this.

Thanks. Will do accordingly in a fresh profile with limited number of add-ons. 

- What are the recommended steps to grab the debug output?

Thank you.

**Comment by mbnuqw** [comment]
@gsbabil, you can open debug info and copy it in Help section in the Sidebery settings page.

**Comment by gsbabil** [comment]
@mbnuqw 

> @gsbabil, you can open debug info and copy it in Help section in the Sidebery settings page.

Thanks you.

I have added a new issue [here](https://github.com/mbnuqw/sidebery/issues/233) with the necessary details. Please have a look. Thanks.

**Comment by Vaelek** [comment]
Coming from TreeTabs, the search is the feature I miss the most. With it not uncommon to have 500 or more tabs across all panels, search is a necessity. Having to click on it matters nothing to me, I just need the functionality. TreeTabs did it wonderfully. Could search title or url, and it highlighted entries as you typed.

**Comment by vurtomatic** [comment]
I'm trying out Sideberry and looking for a way to search open tabs across all browser windows too.

Previously I tried [Saka](https://addons.mozilla.org/en-US/firefox/addon/saka/) which I thought is a nice implementation, echoing @btglr suggestion for something like a Spotlight (Mac) popup.

**Comment by simin75simin** [comment]
use ```o``` from the vimium plugin

**Comment by jlokier** [comment]
> I've tried the % trick in the URL bar and it doesn't seem to find all matching tabs, maybe because the tabs are unloaded:

Using the % trick just now in Firefox 90.0b12, I found that it _did_ find unloaded tabs, but only when they are in the same _container_ as the New Tab page.

I find this annoying in practice because I can rarely remember which container some content was in, so the search is pointless. Searching all containers would be useful. But it explains why Firefox doesn't find tabs even when they are visible.

**Comment by Maki711** [comment]
is this still not implemented ?

**Comment by albino1** [comment]
> is this still not implemented ?

It's in v5, which is technically still in beta but completely functional:

https://github.com/mbnuqw/sidebery/releases

**Comment by matrixik** [comment]
Will this search work like Firefox built-in tab search with only searching titles and urls or like https://addons.mozilla.org/pl/firefox/addon/search-all-tabs/ searching also in the content of all pages?

**Comment by levinotik** [comment]
👍 for search!

**Comment by zesty** [comment]
I have waaay toooooo many tabs for the native % search tabs feature to be useful. Please have a look at the (no longer maintained) "All Tabs Helper" and its separate window usable for searching tabs across all windows. It can search title/url only or also content; temporary sort by title or url+title.

Next level would be a tag cloud based on TF/IDF, but that's probably going to be down to me to make something like that happen.

**Comment by AJolly** [comment]
Came here to check if sideberry would work with tabs across all my windows

**Comment by rublev** [comment]
> Came here to check if sideberry would work with tabs across all my windows

Some alfred-like interface would be great

**Comment by rublev** [comment]
I press ctrl+s and i get a search popup in the top right and in sidebery, but it doesn't auto focus it, unless i was already focused in sidebery first

**Comment by towc** [comment]
hint to users that come here after googling "sidebery search not focusing", I'm having the same issue: press tab and it will focus.

E.g. I bound search to `c-y` as it's the only ctrl binding not set on my ff, and now I can open it and focus with `c-y <tab>`.

I wish I could just do `c-y` and have it auto-focus, but I can 100% believe it's hard to do. To my understanding, the "TabSearch" extension does this by triggering the `_execute_browser_action` which opens the options popup, as if clicking the extension icon in the address bar, which automatically focuses it, and then has the functionality inside of what would normally be the options popup. Sidebery can't do the same trick because our `_execute_browser_action` is already `c-e`, and is tied to toggling the sidebar`, I'm not sure if there's other actions that steal focus. If anyone has come across other methods used by other extensions, I'm sure it would be helpful to point them out!

mbnuqw: maybe a similar trick can be used though: when we press `c-e`, the focus is now probably on the sidebar document, which means we can set the focus to the search bar inside of it? I'd love it if there was a setting to allow `c-e` to automatically focus the search bar inside of the sidebar. It wouldn't be ideal, and maybe there's ways around it (multiple keybinds to open the sidebar, and depending on which was used, search is focused?), but it would be an improvement over `c-y tab` for me

**Comment by SylwesterZarebski** [comment]
> It's in v5, which is technically still in beta but completely functional:
> 
> https://github.com/mbnuqw/sidebery/releases

@albino1 How to use it? I have 5.6.1 version and haven't find any search field or button...

**Comment by albino1** [comment]
> [@albino1](https://github.com/albino1) How to use it? I have 5.6.1 version and haven't find any search field or button...

Have you enabled it in Settings -> Search?  I have it set to `always shown`.

**Comment by SylwesterZarebski** [comment]
> > [@albino1](https://github.com/albino1) How to use it? I have 5.6.1 version and haven't find any search field or button...
> 
> Have you enabled it in Settings -> Search? I have it set to `always shown`.

No, i had default setting - *dynamic*. What this means? It looks like *dynamic* does nothing.
Switching to *always shown* works, thanks!

**Comment by BurntFood** [comment]
> > > [@albino1](https://github.com/albino1) How to use it? I have 5.6.1 version and haven't find any search field or button...
> > 
> > 
> > Have you enabled it in Settings -> Search? I have it set to `always shown`.
> 
> No, i had default setting - _dynamic_. What this means? It looks like _dynamic_ does nothing. Switching to _always shown_ works, thanks!

You have to set a key bind or add the button for it to the navigation bar in sidebery's settings.

**Comment by SylwesterZarebski** [comment]
> > No, i had default setting - _dynamic_. What this means? It looks like _dynamic_ does nothing. Switching to _always shown_ works, thanks!
> 
> You have to set a key bind or add the button for it to the navigation bar in sidebery's settings.

Which bind is suitable for this?
